src/lib.rs

```diff
@@ -1,25 +1,73 @@
 use itertools::*;
 use lazy_static::*;
+use rand::distributions::*;
+use rand::{CryptoRng, Rng};
 use regex::Regex;
 use sha2::{Digest, Sha256};
 use std::collections::HashSet;
+use std::ops::DerefMut;
+use std::sync::{Mutex, MutexGuard};
 use unidecode::unidecode_char;
 use voca_rs::*;
+use Result::{Err, Ok};
 
 lazy_static! {
     ///Special chars that should be filtered out.
     static ref SPECIAL_CHAR: Regex = Regex::new(r#"[!@#$%^&*(){}_<>:;,."'`|+=/~\[\]\\-]"#).unwrap_or_else(|e| panic!("Developer error. Bad regex {:?}", e));
+    static ref ALL_U32: Uniform<u32> = Uniform::new_inclusive(0u32, u32::max_value());
+    //We use this so we don't have to generate the floating numbers and do comparisons on them. It allows us to do 1/2 percent level scaling.
+    static ref ONE_TO_TWO_HUNDRED: Uniform<u8> = Uniform::new_inclusive(1, 200);
+}
+
+const MAX_HASHES_SIZE: usize = 225;
+
+/// Make an index, for the string s considering all tri-grams.
+/// The string will be latinised, lowercased and stripped of special chars before being broken into tri-grams.
+/// The values will be prefixed with partition_id and salt before being hashed.
+/// Each entry in the HasheSet will be truncated to 32 bits and will be encoded as a big endian number.
+/// This function will also add some random entries to the HashSet to not expose how many tri-grams were actually found.
+pub fn generate_hashes_for_string_with_padding<R: Rng + CryptoRng>(
+    s: &str,
+    partition_id: Option<&str>,
+    salt: &[u8],
+    rng: &Mutex<R>,
+) -> Result<HashSet<u32>, String> {
```

Comment:
Does this use a `HashSet` instead of a `Vec` to de-dupe?

Reply:
Yep.